src/main/java/com/yelp/nrtsearch/server/luceneserver/highlights/FastVectorHighlighter.java

```diff
@@ -0,0 +1,141 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.yelp.nrtsearch.server.luceneserver.highlights;
+
+import com.yelp.nrtsearch.server.luceneserver.field.TextBaseFieldDef;
+import java.io.IOException;
+import java.util.Map;
+import org.apache.lucene.document.FieldType;
+import org.apache.lucene.index.IndexReader;
+import org.apache.lucene.search.Query;
+import org.apache.lucene.search.highlight.DefaultEncoder;
+import org.apache.lucene.search.vectorhighlight.FieldQuery;
+import org.apache.lucene.search.vectorhighlight.FragListBuilder;
+import org.apache.lucene.search.vectorhighlight.FragmentsBuilder;
+import org.apache.lucene.search.vectorhighlight.ScoreOrderFragmentsBuilder;
+import org.apache.lucene.search.vectorhighlight.SimpleFragListBuilder;
+import org.apache.lucene.search.vectorhighlight.SimpleFragmentsBuilder;
+import org.apache.lucene.search.vectorhighlight.SingleFragListBuilder;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * The FVH highlighter class for a search query. The highlighted field must be indexed with term
+ * vector with positions and offsets. And the multivalue fields are always fragmented and scored
+ * discretely.
+ */
+public class FastVectorHighlighter implements Highlighter {
+
+  static final String HIGHLIGHTER_NAME = "fast-vector-highlighter";
+  private static final org.apache.lucene.search.vectorhighlight.FastVectorHighlighter
+      FAST_VECTOR_HIGHLIGHTER =
+          new org.apache.lucene.search.vectorhighlight.FastVectorHighlighter();
+  private static final SimpleFragListBuilder SIMPLE_FRAG_LIST_BUILDER = new SimpleFragListBuilder();
+  private static final SingleFragListBuilder SINGLE_FRAG_LIST_BUILDER = new SingleFragListBuilder();
+  private static final SimpleFragmentsBuilder SIMPLE_FRAGMENTS_BUILDER =
+      new SimpleFragmentsBuilder();
+  private static final ScoreOrderFragmentsBuilder SCORE_ORDER_FRAGMENTS_BUILDER =
+      new ScoreOrderFragmentsBuilder();
+  private static final DefaultEncoder DEFAULT_ENCODER = new DefaultEncoder();
+
+  private static final FastVectorHighlighter INSTANCE = new FastVectorHighlighter();
+
+  private static Logger logger = LoggerFactory.getLogger(FastVectorHighlighter.class);
+
+  public static FastVectorHighlighter getInstance() {
+    return INSTANCE;
+  }
+
+  static {
+    // always treat multivalue field discretely
+    SIMPLE_FRAGMENTS_BUILDER.setDiscreteMultiValueHighlighting(true);
+    SCORE_ORDER_FRAGMENTS_BUILDER.setDiscreteMultiValueHighlighting(true);
```

Comment:
Can we skip any changes to the behavior of fast vector highlighter as part of this PR? We would need to check how this would affect existing use-cases.
You can add more context in description on any changes in Fast vector highlighter if they are needed as part of this PR.